src/DiagramGenerator/Image/Storage.php

```diff
@@ -190,4 +178,59 @@ protected function cacheImage($remoteImageUrl, $cachedFilePath)
 
         rename($cachedFilePathTmp, $cachedFilePath);
     }
+
+    private function downloadPieceImages(Config $config)
+    {
+        $pieces = Piece::generateAllPieces();
+
+        $pieceThemeName = $config->getTheme()->getName();
+        $cellSize = $config->getSize()->getCell();
+        @mkdir($this->cacheDirectory.'/'.$pieceThemeName.'/'.$cellSize, 0777, true);
+
+        $handles = [];
+        $fileHandles = [];
+        $multiHandle = curl_multi_init();
+
+        $pieceThemeName = $config->getTheme()->getName();
+        $cellSize = $config->getSize()->getCell();
+
+        foreach ($pieces as $piece) {
+            $pieceUrl = $this->generatePieceUrl($piece, $config);
+            $handles[$piece->getShortName()] = curl_init($pieceUrl);
+            $fileUrl = $this->getCachedPieceFilePath($pieceThemeName, $cellSize, $piece->getShortName());
+            $fileHandles[$piece->getShortName()] = fopen($fileUrl, 'wb');
+        }
+
+        foreach($handles as $key => $handle) {
+            curl_setopt($handle, CURLOPT_FILE, $fileHandles[$key]);
```

Comment:
Opening the filehandles all at once will delete any files that are already in the process of being downloaded by a different thread. Should we instead write to a filehandle with a random name, and then move them to the final location atomically once complete?